test/jdk/java/nio/file/Files/InterruptCopy.java

```diff
@@ -101,19 +119,30 @@ public void run() {
 
             // copy source to target via task in thread pool, interrupting it after
             // a delay using cancel(true)
+            final CountDownLatch clatch = new CountDownLatch(2);
             Future<Void> result = pool.submit(new Callable<Void>() {
                 public Void call() throws IOException {
                     System.out.println("Copying file...");
+                    clatch.countDown();
+                    try {
+                        clatch.await();
+                    } catch (InterruptedException ignored) {
+                    }
                     Files.copy(source, target, ExtendedCopyOption.INTERRUPTIBLE,
                         StandardCopyOption.REPLACE_EXISTING);
                     return null;
                 }
             });
-            Thread.sleep(DELAY_IN_MS);
+            clatch.countDown();
+            clatch.await();
+            Thread.sleep(CANCEL_DELAY_IN_MS);
             boolean cancelled = result.cancel(true);
-            if (!cancelled)
+            if (cancelled)
+                System.out.println("Copy cancelled.");
+            else {
                 result.get();
-            System.out.println("Copy cancelled.");
+                throw new RuntimeException("Copy was not cancelled");
```

Comment:
Should `DURATION_MAX_IN_MS` be taken into account here too to decide whether to throw or not?

Reply:
I don't know. It seems like cancellation ought to be a bit more predictable than interruption. I am not sure how `DURATION_MAX_IN_MS` was determined. If there were no "hiccups" in the threads starting, then with the correct timing I am not sure that such a threshold is even needed. The execution time for the copy that I have observed is between 150ms (macOS) and 400ms (Windows) which is a lot less than 5s, so there would have to be some slowdown for this threshold even to be reached.

Reply:
Not wishing to labour the discussion on the change, which will in the main deal with the issue, but is it not the case with the current set of Intermittent failures that there is this slowdown, i.e. the copy is taking longer than the threshold of 5 seconds hence the RuntimeException with Copy was not Interrupted?

As such,

if (duration > DURATION_MAX_IN_MS)
                    throw new RuntimeException("Copy was not interrupted");

is triggering a failure. 
You would expect if a typical copy duration is 400msecs that if the duration is taking a lot longer that the interrupt has a greater probability to happen.

the rationale for this logic is not clear - so if the duration is less than the copy threshold then test proceeds as if an interrupt has occurred?

the success and failure semantics for the test are not absolute. As such, maybe there's an opportunity to be more exact with test semantics e.g. retrying to interrupt the copy if it didn't happen. 

as an aside it would appear that invoking Thread.interrupted(); in the main thread seems redundant ?